### PR TITLE
Inner-Product Verification with Multiexponentiation | Inner-Product Protocol for Non-powers of 2

### PR DESCRIPTION
**Pull request by suyash67**
1. Faster inner product proof verification using a single multi-exponentiation as described in Section 3.1 of the [Bulletproofs](https://eprint.iacr.org/2017/1066.pdf) paper. Improves the inner product verification by **~30%** 

2. Protocols similar in spirit to Bulletproofs require support for inner product protocol with secret vector sizes a non-power of 2. I have added a test demonstrating the same. The idea is simple: 
2.1 Append secret vectors with 0's to make them the next power of 2
2.2 Extend the original base vectors to make them too the next power of 2
2.3 Compute vector exponentiations with a condition to check any 0's. This ensures minimum overhead computational cost due to modification of vector sizes to the next power of 2.

### CODE REVIEW

**Comment by omershlo** [comment]
looks fantastic !

File: src/proofs/inner_product.rs [hunk 1]
```diff
@@ -324,6 +435,177 @@ mod tests {
...
+        //     if x.1 != BigInt::zero() {
+        //         // Mult and Add only if the element cL[i] is not zero
+        //         let cL_i_fe: FE = ECScalar::from(&x.1);
+        //         let cL_i_fe_g_i: GE = x.0 * &cL_i_fe;
+        //         acc.add_point(&cL_i_fe_g_i.get_element())
+        //     } else {
+        //         // move on otherwise
+        //         acc
+        //     }
+        // });
```
Comment (omershlo) [inline src/proofs/inner_product.rs:586]: please remove

File: src/proofs/inner_product.rs [hunk 2]
```diff
@@ -324,6 +435,177 @@ mod tests {
...
+        //         rand.to_big_int()
+        //     })
+        //     .collect();
+
+        // let b: Vec<_> = (0..n)
+        //     .map(|_| {
+        //         let rand: FE = ECScalar::new_random();
+        //         rand.to_big_int()
+        //     })
+        //     .collect();
```
Comment (omershlo) [inline src/proofs/inner_product.rs:554]: remove

File: src/proofs/inner_product.rs [hunk 3]
```diff
@@ -218,6 +236,99 @@ impl InnerProductArg {
...
+
+            let x = HSha256::create_hash_from_ge(&[&Li, &Ri, &ux]);
+            let x_bn = x.to_big_int();
+            let x_inv_fe = x.invert();
+            let x_inv_bn = x_inv_fe.to_big_int();
+            let x_sq_bn = BigInt::mod_mul(&x_bn, &x_bn, &order);
+            let x_inv_sq_bn =
+                BigInt::mod_mul(&x_inv_fe.to_big_int(), &x_inv_fe.to_big_int(), &order);
+            // let x_sq_fe: FE = ECScalar::from(&x_sq_bn);
+            // let x_inv_sq_fe: FE = ECScalar::from(&x_inv_sq_bn);
```
Comment (omershlo) [inline src/proofs/inner_product.rs:276]: remove

File: src/proofs/inner_product.rs [hunk 4]
```diff
@@ -218,6 +236,99 @@ impl InnerProductArg {
...
+            x_inv_sq_vec.push(x_inv_sq_bn.clone());
+            minus_x_sq_vec.push(BigInt::mod_sub(&BigInt::zero(), &x_sq_bn, &order));
+            minus_x_inv_sq_vec.push(BigInt::mod_sub(&BigInt::zero(), &x_inv_sq_bn, &order));
+            allinv = allinv * x_inv_bn;
+        }
+
+        let mut s: Vec<BigInt> = Vec::with_capacity(n);
+        s.push(allinv);
+        for i in 1..n {
+            let lg_i = (32 - 1 - (i as u32).leading_zeros()) as usize;
```
Comment (omershlo) [inline src/proofs/inner_product.rs:288]: what is the meaning of the number 32 ?

File: src/proofs/inner_product.rs [hunk 5]
```diff
@@ -218,6 +236,99 @@ impl InnerProductArg {
...
+            let k = 1 << lg_i;
+            // The challenges are stored in "creation order" as [x_k,...,x_1],
+            // so u_{lg(i)+1} = is indexed by (lg_n-1) - lg_i
+            let x_lg_i_sq = x_sq_vec[(lg_n - 1) - lg_i].clone();
+            s.push(s[i - k].clone() * x_lg_i_sq);
+        }
```
Comment (omershlo) [inline src/proofs/inner_product.rs:294]: could you explain in short what is the logic of computing the `s` vector? 
it is a bit hard to follow

Comment (suyash67) [reply src/proofs/inner_product.rs:292]: ## A bit of theory

In the usual inner product proof verification, in the _n_-th round, we compute new base vectors computed as follows: 

<a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;\begin{align*}&space;\textbf{g}^{\prime}&space;&=&space;\textbf{g}^{x^{-1}}_{[:n^{\prime}]}&space;\&space;\circ&space;\&space;\textbf{g}^{x}_{[n^{\prime}:]}&space;\in&space;\mathbb{G}^{\frac{n}{2}}&space;\\&space;\textbf{h}^{\prime}&space;&=&space;\textbf{h}^{x}_{[:n^{\prime}]}&space;\&space;\circ&space;\&space;\textbf{h}^{x^{-1}}_{[n^{\prime}:]}&space;\in&space;\mathbb{G}^{\frac{n}{2}}&space;\end{align*}" target="_blank"></a>

Thus, the base vectors used in each round depend only on the challenges  <a href="https://www.codecogs.com/eqnedit.php?latex=x_1,&space;x_2,&space;\dots,&space;x_{\lceil&space;\text{log}_2n&space;\rceil}" target="_blank"></a> used in each round. In _i_-th round, the verifier computes <a href="https://www.codecogs.com/eqnedit.php?latex=P_{i&plus;1}" target="_blank"></a> which would be used in the next round. It is computed as:

<a href="https://www.codecogs.com/eqnedit.php?latex=P_{i&plus;1}&space;=&space;L_{i}^{x_i^{2}}&space;P_{i}&space;R_{i}^{x_i^{-2}}" target="_blank"></a>

In the last round, the verifier checks the following equality:

<a href="https://www.codecogs.com/eqnedit.php?latex=P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}&space;=&space;g^a&space;h^bu^{a&space;\cdot&space;b}" target="_blank"></a>

where <a href="https://www.codecogs.com/eqnedit.php?latex=g,h" target="_blank"></a> are generators used in the last round of the protocol.

For a single multi-exponentiation verification, a verifier just needs to just check the last verification equation. For that, a verifier will need the following quantities:

<a href="https://www.codecogs.com/eqnedit.php?latex=g,h,&space;\text{&space;and&space;}&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}" target="_blank"></a>

Computing <a href="https://www.codecogs.com/eqnedit.php?latex=P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}" target="_blank"></a> is easy. We need to do the following:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;P_{\lceil&space;\text{log}_2n&space;\rceil}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;\left(&space;L_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^2)}&space;\cdot&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^{-2})}&space;\right)&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;L_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^2)}&space;\bigg(&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;\bigg)&space;R_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^{-2})}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;\left(&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;L_{j}^{(x_j^2)}&space;\right)&space;\cdot&space;P&space;\cdot&space;\left(&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;R_{j}^{(x_j^{-2})}&space;\right)&space;\end{align}" target="_blank"><img src="https://latex.codecogs.com/gif.latex?\begin{align*}&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;&plus;&space;1}&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;P_{\lceil&space;\text{log}_2n&space;\rceil}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;\left(&space;L_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^2)}&space;\cdot&space;P_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^{-2})}&space;\right)&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;L_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^2)}&space;\cdot&space;L_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^2)}&space;\bigg(&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;.&space;\&space;\bigg)&space;R_{\lceil&space;\text{log}_2n&space;\rceil&space;-1}^{(x_j^{-2})}&space;\cdot&space;R_{\lceil&space;\text{log}_2n&space;\rceil}^{(x_j^{-2})}&space;\\&space;&=&space;\left(&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;L_{j}^{(x_j^2)}&space;\right)&space;\cdot&space;P&space;\cdot&space;\left(&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;R_{j}^{(x_j^{-2})}&space;\right)&space;\end{align}" title="\begin{align*} P_{\lceil \text{log}_2n \rceil + 1} &= L_{\lceil \text{log}_2n \rceil}^{(x_j^2)} \cdot P_{\lceil \text{log}_2n \rceil} \cdot R_{\lceil \text{log}_2n \rceil}^{(x_j^{-2})} \\ &= L_{\lceil \text{log}_2n \rceil}^{(x_j^2)} \cdot \left( L_{\lceil \text{log}_2n \rceil -1}^{(x_j^2)} \cdot P_{\lceil \text{log}_2n \rceil -1} \cdot R_{\lceil \text{log}_2n \rceil -1}^{(x_j^{-2})} \right) \cdot R_{\lceil \text{log}_2n \rceil}^{(x_j^{-2})} \\ &= L_{\lceil \text{log}_2n \rceil}^{(x_j^2)} \cdot L_{\lceil \text{log}_2n \rceil -1}^{(x_j^2)} \bigg( \ . \ . \ . \ . \ . \ . \ \bigg) R_{\lceil \text{log}_2n \rceil -1}^{(x_j^{-2})} \cdot R_{\lceil \text{log}_2n \rceil}^{(x_j^{-2})} \\ &= \left( \prod_{j=1}^{\lceil \text{log}_2n \rceil} L_{j}^{(x_j^2)} \right) \cdot P \cdot \left( \prod_{j=1}^{\lceil \text{log}_2n \rceil} R_{j}^{(x_j^{-2})} \right) \end{align}" /></a>

We can compute <a href="https://www.codecogs.com/eqnedit.php?latex=g,h" target="_blank"></a> by unrolling the recursion and write them in the form:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;g&space;&=&space;\prod_{i&space;=&space;1}^{n}&space;g_i^{s_i}&space;\in&space;\mathbb{G}&space;\\&space;h&space;&=&space;\prod_{i&space;=&space;1}^{n}&space;h_i^{s_i^{-1}}&space;\in&space;\mathbb{G}&space;\end{align*}" target="_blank"></a>

Now to compute `s` vector <a href="https://www.codecogs.com/eqnedit.php?latex=(s_1,&space;s_2,&space;\dots,&space;s_n)" target="_blank"></a>, let's see it through an example for _n=8_.

CodeCogsEqn

CodeCogsEqn (3)

NOTE that I intended to write <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;x_j^{-1}" target="_blank"></a> but instead mistakenly wrote <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;-x_j" target="_blank"></a>. Apologies for the same.

It's pretty clear why the exponents of <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;g_i,&space;h_i" target="_blank"></a> are the inverses of each other for all <a href="https://www.codecogs.com/eqnedit.php?latex=\inline&space;i=1,2,\dots,&space;n." target="_blank"></a>. There is a pattern in the powers of triplet <a href="https://www.codecogs.com/eqnedit.php?latex=(x_1,x_2,x_3)" target="_blank"></a> such that

<a href="https://www.codecogs.com/eqnedit.php?latex=s_i&space;=&space;\prod_{j=1}^{\lceil&space;\text{log}_2n&space;\rceil}&space;x_j^{b_{i,j}}&space;\text{&space;where&space;}&space;b_{i,j}&space;=&space;\begin{cases}&space;1&space;&&space;\text{if&space;}&space;j\text{-th&space;bit&space;of&space;}&space;i-1&space;\text{&space;is&space;}&space;1&space;\\&space;-1&space;&&space;\text{otherwise}&space;\end{cases}" target="_blank"></a>

This part is a bit confusing but when we see our example again, it is much easy to understand.

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;i&space;=&space;1&space;&&space;\implies&space;\&space;(b_{1,1},&space;b_{1,2},&space;b_{1,3})&space;=&space;(-1,&space;-1,&space;-1)&space;&&space;\equiv&space;\&space;000&space;\\&space;i&space;=&space;2&space;&&space;\implies&space;\&space;(b_{2,1},&space;b_{2,2},&space;b_{2,3})&space;=&space;(-1,&space;-1,&space;1)&space;&&space;\equiv&space;\&space;001&space;\\&space;i&space;=&space;3&space;&&space;\implies&space;\&space;(b_{3,1},&space;b_{3,2},&space;b_{3,3})&space;=&space;(-1,&space;1,&space;-1)&space;&&space;\equiv&space;\&space;010&space;\\&space;\vdots&space;&&space;&&space;\\&space;i&space;=&space;8&space;&&space;\implies&space;\&space;(b_{8,1},&space;b_{8,2},&space;b_{8,3})&space;=&space;(1,&space;1,&space;1)&space;&&space;\equiv&space;\&space;111&space;\\&space;\end{align}" target="_blank"></a>

Therefore, for given <a href="https://www.codecogs.com/eqnedit.php?latex=(i,j)" target="_blank"></a>, we can easily compute vector component `s_i`.

## Implementation Idea

First we compute <a href="https://www.codecogs.com/eqnedit.php?latex=s_1&space;=&space;x_k^{-1}&space;\cdot&space;x_{k-1}^{-1}&space;\dots&space;x_2^{-1}&space;\cdot&space;x_1^{-1}&space;\text{&space;where&space;}&space;k=\lceil&space;\text{log}_2n&space;\rceil" target="_blank"></a>. Note the order of the challenges is reverse since challenges are stored in order of creation as <a href="https://www.codecogs.com/eqnedit.php?latex=[x_k,&space;x_{k-1},&space;\dots,&space;x_1]" target="_blank"></a>.

Now, we for next values in `s` vector:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;s_2&space;&=&space;s_1&space;\cdot&space;x_3^{2}&space;\\&space;s_3&space;&=&space;s_1&space;\cdot&space;x_{2}^{2}&space;\\&space;s_4&space;&=&space;s_2&space;\cdot&space;x_{2}^2&space;\\&space;\vdots&space;\end{align}" target="_blank"></a>.

The logic is that for computing <a href="https://www.codecogs.com/eqnedit.php?latex=s_i" target="_blank"></a> we need some <a href="https://www.codecogs.com/eqnedit.php?latex=s_{i-k}" target="_blank"></a> in which we just need to flip _one_ bit to get the desired configuration of <a href="https://www.codecogs.com/eqnedit.php?latex=s_i" target="_blank"></a>. Let me know if there are any issues in my explanation.

Comment (suyash67) [reply src/proofs/inner_product.rs:292]: Please note there could also be an alternate way of computing <a href="https://www.codecogs.com/eqnedit.php?latex=s_i" target="_blank"></a>. Let us discuss if we can do it an easier way.

File: src/proofs/inner_product.rs (hunk 4 above)
Comment (suyash67) [reply src/proofs/inner_product.rs:286]: As explained in my answer to your next question, once we compute <a href="https://www.codecogs.com/eqnedit.php?latex=s_{1}" target="_blank"></a>, we need to compute <a href="https://www.codecogs.com/eqnedit.php?latex=s_{2},&space;s_3,&space;\dots" target="_blank"></a> according to the bits in <a href="https://www.codecogs.com/eqnedit.php?latex=(i-1)" target="_blank"></a> for <a href="https://www.codecogs.com/eqnedit.php?latex=i&space;=&space;2,3,\dots" target="_blank"></a>.

Here, I am iterating (i-1) as 1,2,...(n-1). 32 just means that I am assuming that 

<a href="https://www.codecogs.com/eqnedit.php?latex=\text{log}_2n&space;\le&space;32&space;\implies&space;n&space;\le&space;2^{32}&space;\approx&space;4.29&space;\text{&space;Billion}" target="_blank"></a>

I should have mentioned it in a comment in the code. Hope this helps.

File: src/proofs/inner_product.rs (hunk 5 above)
Comment (omershlo) [reply src/proofs/inner_product.rs:292]: Got it! thank you for the thorough explanation! neat idea !

File: src/proofs/inner_product.rs (hunk 4 above)
Comment (omershlo) [reply src/proofs/inner_product.rs:286]: ok, so why you assume 32 and not take as a parameter ? 
- I think you should add a check for n and take the next number as this parameter
- or , if you are convinced that it must be 32 you should add a bound check, throwing error in case n is too large and exit.

Comment (suyash67) [reply src/proofs/inner_product.rs:286]: You are right. We don't need 32 necessarily. We can take it as a parameter. I think it would be a bit confusing for a user if we let the user decide this. Instead I can make it 64 (which is far more than enough according to me) and add a check and throw an error if sizes greater than 64 are encountered.

Comment (omershlo) [reply src/proofs/inner_product.rs:286]: can't it be a function of `n` ?

Comment (suyash67) [reply src/proofs/inner_product.rs:286]: Oh right. I am just using it in computing log_2(i). So it should work even when it depends on n. I will try setting it just to log_2(n).

Comment (suyash67) [reply src/proofs/inner_product.rs:286]: So a constant like 32 is necessary because I am using it to compute floor of log_2 of integers 1,2,..,(n-1). The way I am doing this is as follows:

<a href="https://www.codecogs.com/eqnedit.php?latex=\begin{align*}&space;&&space;i=13&space;\text{&space;as&space;u32&space;}&space;\implies&space;\&space;\overbrace{&space;000\dots&space;00}^{28}1101&space;\\&space;&&space;\qquad&space;\lfloor&space;\text{log}_2(i)&space;\rfloor&space;=&space;32-1-28&space;=&space;3&space;\end{align*}" target="_blank"></a>

Now we could also use 64 instead. But it necessarily has to be 32 (because of u32) or 64 (for u64). I think I will just add a check for <a href="https://www.codecogs.com/eqnedit.php?latex=n&space;>&space;2^{64}" target="_blank">&space;2^{64}" title="n > 2^{64}" /></a>.

Comment (suyash67) [reply src/proofs/inner_product.rs]: I have changed it to support 64. Please check.

Comment (omershlo) [reply src/proofs/inner_product.rs]: What about using BigInt for it? 
there's a function to get the bit length of a BigInt.

Comment (suyash67) [reply src/proofs/inner_product.rs]: The length `n` we compute in verification has the datatype `usize`.
So anyway we cannot have `n > 2^{64}` (for 64-bit targets). Are you suggesting that I try setting `n` itself to BigInt and change the scalar computations accordingly?

Comment (omershlo) [reply src/proofs/inner_product.rs]: you make a good point. lets keep the 64bit and not overkill. I will merge.
